test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1307 @@
+
+		if (odp_pktio_config(pktio->handle, &pktio_config) < 0) {
+			ODPH_ERR("Error configuring packet I/O extra options (%s)\n", pktio->name);
+			return false;
+		}
+
+		(void)odp_pktio_mac_addr(pktio->handle, &pktio->src_mac, sizeof(pktio->src_mac));
+
+		if (odp_pktio_start(pktio->handle) < 0) {
+			ODPH_ERR("Error starting packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static odp_bool_t setup_fwd_table(prog_config_t *config)
+{
+	fwd_entry_t *fwd_e;
+
+	config->fwd_tbl = odph_iplookup_table_create(PROG_NAME, 0U, 0U, sizeof(fwd_entry_t *));
+
+	if (config->fwd_tbl == NULL) {
+		ODPH_ERR("Error creating forwarding table\n");
+		return false;
+	}
+
+	for (uint32_t i = 0U; i < config->num_fwds; ++i) {
+		fwd_e = &config->fwd_entries[i];
+
+		if (odph_iplookup_table_put_value(config->fwd_tbl, &fwd_e->prefix, &fwd_e) < 0) {
+			ODPH_ERR("Error populating forwarding table\n");
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static inline odp_ipsec_sa_t *get_in_sa(odp_packet_t pkt)
+{
+	odph_esphdr_t esp;
+	uint32_t spi;
+
+	if (!odp_packet_has_ipsec(pkt))
+		return NULL;
+
+	if (odp_packet_copy_to_mem(pkt, odp_packet_l4_offset(pkt), ODPH_ESPHDR_LEN, &esp) < 0)
+		return NULL;
+
+	spi = odp_be_to_cpu_32(esp.spi);
+
+	return spi <= UINT16_MAX ? spi_to_sa_map[DIR_IN][spi] : NULL;
+}
+
+static inline odp_bool_t process_ipsec_in(odp_packet_t pkt, const odp_ipsec_sa_t *sa)
+{
+	odp_ipsec_in_param_t param;
+
+	memset(&param, 0, sizeof(param));
+	param.num_sa = 1;
+	param.sa = sa;
+	/* IPsec in/out need to be identified somehow, so use user_ptr for this. */
+	odp_packet_user_ptr_set(pkt, NULL);
+
+	return odp_ipsec_in_enq(&pkt, 1, &param) == 1;
+}
+
+static inline odp_ipsec_sa_t *get_out_sa(odp_packet_t pkt)
+{
+	odph_udphdr_t udp;
+	uint16_t dst_port;
+
+	if (!odp_packet_has_udp(pkt))
+		return NULL;
+
+	if (odp_packet_copy_to_mem(pkt, odp_packet_l4_offset(pkt), ODPH_UDPHDR_LEN, &udp) < 0)
+		return NULL;
+
+	dst_port = odp_be_to_cpu_16(udp.dst_port);
+
+	return dst_port ? spi_to_sa_map[DIR_OUT][dst_port] : NULL;
+}
+
+static inline odp_bool_t process_ipsec_out(odp_packet_t pkt, const odp_ipsec_sa_t *sa)
+{
+	odp_ipsec_out_param_t param;
+
+	memset(&param, 0, sizeof(param));
+	param.num_sa = 1;
+	param.sa = sa;
+	/* IPsec in/out need to be identified somehow, so use user_ptr for this. */
+	odp_packet_user_ptr_set(pkt, sa);
+
+	return odp_ipsec_out_enq(&pkt, 1, &param) == 1;
+}
+
+static inline const pktio_t *lookup_and_apply(odp_packet_t pkt, odph_table_t fwd_tbl,
+					      odp_bool_t is_eth_mod)
+{
+	const uint32_t l3_off = odp_packet_l3_offset(pkt);
+	odph_ipv4hdr_t ipv4;
+	uint32_t ip, l2_off;
+	fwd_entry_t *fwd;
+	odph_ethhdr_t eth;
+
+	if (odp_packet_copy_to_mem(pkt, l3_off, ODPH_IPV4HDR_LEN, &ipv4) < 0)
+		return NULL;
+
+	ip = odp_be_to_cpu_32(ipv4.dst_addr);
+
+	if (odph_iplookup_table_get_value(fwd_tbl, &ip, &fwd, 0U) < 0 || fwd == NULL)
+		return NULL;
+
+	if (is_eth_mod) {
```

Comment:
The is_eth_mod parameter here and the related passing of it through the call chain is a bit messy. And I do not think it buys you much. It basically tries to optimize for the plain forwarding case (without IPsec), which is not the focus here anyway, and does not really help much even in that case.

I would like to get rid of the parameter (or change it to ethertype parameter if really needed) and just write IPv4 ethertype in packets in all cases. If you are worried about putting an incorrect ethertype in the packet (e.g. IPv6 received), there is the same problem even after IPsec processing (IPv6 packets may come out of an IPv4 tunnel). You could check and drop all non-IPv4 traffic if you wanted but this is just a simple perf test tool, so maybe it is not needed.

If is_eth_mod goes away, then you would check if l3_off is exactly ODPH_ETHHDR_LEN and avoid the push_head in that case. If l3_off was not exactly ODPH_ETHHDR_LEN in the "is_eth_mod" case, the current code would not work anyway since extra bytes before l2 offset would be sent to the wire.

Reply:
Simplified in v4.